### ISSUE TITLE
Getting: segmentation fault on local env. Perhaps my installation is wrong?

### ISSUE THREAD

**Issue by pmochine** [phpredis/phpredis#1835]
So I have switched from predis to phpredis. However, I get segmentation fault errors :( I created a [thread on stackoverflow](https://stackoverflow.com/questions/63651799/changed-predis-to-phpredis-getting-segmentation-fault) but had no luck.

### I'm seeing this behaviour on
- OS: 
ProductName:	Mac OS X
ProductVersion:	10.15.6
BuildVersion:	19G73
- Redis: 6.0.6
- PHP: 7.4.9
- phpredis: 5.3.1

### Steps to reproduce, backtrace or example script

First of all. My first question is: I have installed Redis with brew month ago. Now I've installed phpredis with `pecl redis`. Do you think this could be a problem?

I've created a simple test for Redis, but it, unfortunately, fails with `[1]    27996 segmentation fault`:

    // Config in App.php
    // Had to rename Redis, because of the use of phpredis!
    'RedisManager' => Illuminate\Support\Facades\Redis::class,

    //Test unit
    use Illuminate\Support\Facades\Redis;

    /** @test */
    public function it_can_set_a_dummy_value_in_redis()
    {
        $value = 'Hello';

        Redis::set($this->cacheKey, $value, 'EX', 60);

        $this->assertEquals($value, Redis::get($this->cacheKey));
    }

It fails when the command `set` is fired. 

I went through the code up to the point: `Illuminate\Redis\Connections\PhpRedisConnection.php@command`

I checked what the object-value of `$this->client` is:

    Redis {#2278
      isConnected: true
      host: "127.0.0.1"
      port: 6379
      auth: null
      mode: ATOMIC
      dbNum: 0
      timeout: 0.0
      lastError: null
      persistentId: null
      options: {
        TCP_KEEPALIVE: 0
        READ_TIMEOUT: 0.0
        COMPRESSION: NONE
        SERIALIZER: NONE
        PREFIX: "local_database_"
        SCAN: NORETRY
      }
    }

I check the connection with:

     dd($this->client->ping()); // it's true

However, it fails at the point:

    return parent::command($method, $parameters); //set is called

Even though it is in a try-catch-block, it won't be shown there...

The logs show nothing:
- nginx: 0 errors
- php-fpm: 0 errors
- redis: 

     2867:M 29 Aug 2020 22:46:03.142 * Background saving started by pid 27498
     27498:C 29 Aug 2020 22:46:03.145 * DB saved on disk
     2867:M 29 Aug 2020 22:46:03.306 * Background saving terminated with success

So even Redis is working correctly.

I don't know what is going on. When I change to predis, everything works fine!

The only hint I have got is [this thread](https://stackoverflow.com/questions/23569835/why-should-i-close-or-keep-redis-connections-open). But since my `ping()` is ponged, it should work?

Any ideas how I can fix my problem? And how can I debug this? There are no logs for this error...

Edit: 

    php -r "if (new Redis() == true){ echo \"\r\n OK \r\n\"; }"

Gives me also an "OK"

### I've checked
- [x] There is no similar issue from other users (there are tons but I couldnt find my specific issue)
- [ ] Issue isn't fixed in `develop` branch

## Updates:

I tried a simple test file and it works. So something else must be happening!

```
<?php

$redis = new Redis();

$redis->connect('127.0.0.1', 6379);
$redis->setEx('delete-key', 60, 'some-value'); //works
```

**Comment by yatsukhnenko** [comment]
@pmochine could you provide backtrace for segmentation fault?

**Comment by pmochine** [comment]
@yatsukhnenko Could send me a link, how I can backtrace on a mac? 
I found this https://bugs.php.net/bugs-generating-backtrace.php

However, since I'm using laravel, I'm not quite sure how to do it. Usually, I just start the unit test with `phpunit --filter RedisTest`. But I see examples that I need to do it like: `gdb --args php yourscript.php`. Since my test fails in a framework, I cannot simply call the php file.


### Update
Installed gdb with brew and [this tutorial](https://sourceware.org/gdb/wiki/BuildingOnDarwin).

However, when I run this command: `gdb --args php /Users/someuser/app/vendor/bin/phpunit /Users/someuser/app/Unit/Redis/RedisTest.php`

I get the message:

```
GNU gdb (GDB) 9.2
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-apple-darwin19.5.0".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from php...
(No debugging symbols found in php)
(gdb) r
Starting program: /usr/local/bin/php /Users/someuser/app/vendor/bin/phpunit /Users/someuser/app/tests/Unit/Redis/RedisTest.php
[New Thread 0x2503 of process 1587]
[New Thread 0x2203 of process 1587]
warning: unhandled dyld version (16)
PHPUnit 9.2.5 by Sebastian Bergmann and contributors.

.
Thread 2 received signal SIGSEGV, Segmentation fault.
0x0000000103ec7184 in ?? ()
(gdb) 

```

But my guess ist that this won't help you much right?


### Update 2

Now I have installed xdebug and this is the stack for php before it fails


    XDEBUG_CONFIG="idkey=sublime.xdebug" phpunit --filter RedisTest

```
[0] file:///Users/someuser/app/vendor/laravel/framework/src/Illuminate/Redis/Connections/Connection.php:116, Illuminate\Redis\Connections\PhpRedisConnection->command()
[1] file:///Users/someuser/app/vendor/laravel/framework/src/Illuminate/Redis/Connections/PhpRedisConnection.php:547, Illuminate\Redis\Connections\PhpRedisConnection->command()
[2] file:///Users/someuser/app/vendor/laravel/framework/src/Illuminate/Redis/Connections/PhpRedisConnection.php:86, Illuminate\Redis\Connections\PhpRedisConnection->set()
[3] file:///Users/someuser/app/vendor/laravel/framework/src/Illuminate/Redis/RedisManager.php:263, Illuminate\Redis\RedisManager->__call()
[4] file:///Users/someuser/app/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php:261, Illuminate\Redis\RedisManager->set()
[5] file:///Users/someuser/app/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php:261, Illuminate\Support\Facades\Facade::__callStatic()
[6] file:///Users/someuser/app/tests/Unit/Redis/RedisTest.php:27, Illuminate\Support\Facades\Facade::set()
[7] file:///Users/someuser/app/tests/Unit/Redis/RedisTest.php:27, Tests\Unit\Redis\RedisTest->it_can_set_a_dummy_value_in_redis()
[8] file:///Users/someuser/app/vendor/phpunit/phpunit/src/Framework/TestCase.php:1413, Tests\Unit\Redis\RedisTest->runTest()
[9] file:///Users/someuser/app/vendor/phpunit/phpunit/src/Framework/TestCase.php:1030, Tests\Unit\Redis\RedisTest->runBare()
[10] file:///Users/someuser/app/vendor/phpunit/phpunit/src/Framework/TestResult.php:692, PHPUnit\Framework\TestResult->run()
[11] file:///Users/someuser/app/vendor/phpunit/phpunit/src/Framework/TestCase.php:771, Tests\Unit\Redis\RedisTest->run()
[12] file:///Users/someuser/app/vendor/phpunit/phpunit/src/Framework/TestSuite.php:638, PHPUnit\Framework\TestSuite->run()
[13] file:///Users/someuser/app/vendor/phpunit/phpunit/src/Framework/TestSuite.php:638, PHPUnit\Framework\TestSuite->run()
[14] file:///Users/someuser/app/vendor/phpunit/phpunit/src/Framework/TestSuite.php:638, PHPUnit\Framework\TestSuite->run()
[15] file:///Users/someuser/app/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:657, PHPUnit\TextUI\TestRunner->run()
[16] file:///Users/someuser/.composer/vendor/phpunit/phpunit/src/TextUI/Command.php:107, PHPUnit\TextUI\Command->run()
[17] file:///Users/someuser/.composer/vendor/phpunit/phpunit/src/TextUI/Command.php:67, PHPUnit\TextUI\Command::main()
[18] file:///Users/someuser/.composer/vendor/phpunit/phpunit/phpunit:61, {main}()
```

It fails exactly what I told you at the beginning of this post. [The set command](https://github.com/laravel/framework/blob/c62385a23c639742b3b74a4a78640da25e6b782b/src/Illuminate/Redis/Connections/PhpRedisConnection.php#L77) and than of [the abstractor](https://github.com/laravel/framework/blob/c62385a23c639742b3b74a4a78640da25e6b782b/src/Illuminate/Redis/Connections/Connection.php#L96)

**Comment by tradzero** [comment]
same problem when use laravel's Redis Facade

**Comment by meanevo** [comment]
+1 for having this issue.
env: phpredis 5.3.1 on macos 10.15.7

As @pmochine mentioned, I followed the Laravel facade code and find out that the way it used to calling phpredis.
`Redis::set('test', 123456, 'EX', '60')` is being translated into `$client->set('test', 123456, [null, 'EX' => 60])`, and calling this method manually also rises an segfault.
**Note the third argument's first element, is being passed as null**

Then I compiled the ext from source and re-debug the code with lldb, I got an EXC_BAD_ACCESS.
![image](https://user-images.githubusercontent.com/1898273/95846378-f193cb00-0d7d-11eb-995b-bc12fac9b045.png)
> Related code: https://github.com/phpredis/phpredis/blob/5.3.1/redis_commands.c#L1352

Since I'm unfamiliar with clang ;( I can only guess it maybe something todo with the var-"len" being null. 
> https://github.com/phpredis/phpredis/blob/5.3.1/common.h#L219

---
There may have some (dirty) tricks to mitigate this issue temporarily (in Laravel)
1. Using predis as redis driver by adding REDIS_CLIENT=predis to .env;
2. By adding an extra arg(empty string as $flag) when using facade, eg. Redis::set('test', 123456, 'EX', '60', **''**)
3. Calling phpredis library directly, or maybe Redis::setex('test', 60, 123456)
3. Since the issue is only observed on macos, maybe moving development env to container/Docker will help.
